core/src/main/scala/kafka/utils/CoreUtils.scala

```diff
@@ -242,31 +241,6 @@ object CoreUtils {
     properties
   }
 
-  /**
-   * Atomic `getOrElseUpdate` for concurrent maps. This is optimized for the case where
-   * keys often exist in the map, avoiding the need to create a new value. `createValue`
-   * may be invoked more than once if multiple threads attempt to insert a key at the same
-   * time, but the same inserted value will be returned to all threads.
-   *
-   * In Scala 2.12, `ConcurrentMap.getOrElse` has the same behaviour as this method, but JConcurrentMapWrapper that
-   * wraps Java maps does not.
-   */
-  def atomicGetOrUpdate[K, V](map: concurrent.Map[K, V], key: K, createValue: => V): V = {
```

Comment:
Can we double check that this comment is accurate for the various `concurrent.Map` implementations?

Reply:
I have reviewed the implementation in the Scala source code.

We could remove `atomicGetOrUpdate()` in Scala 2.13, as Scala now provides a very similar implementation for `getOrElseUpdate()`.

In Scala 2.12, if you convert a ConcurrentHashMap using `scala.collection.JavaConverters.asScala`, you will indeed get a less reliable version of `getOrElseUpdate` that does not handle race conditions or atomic operations properly. This is the less robust version, as it doesn't ensure atomic updates.

2.13
```scala
  /**
    * Replaces the entry for the given key only if it was previously mapped
    * to some value.
    *
    * $atomicop
    *
    * @param k   key for which the entry should be replaced
    * @param v   value to be associated with the specified key
    * @return    `Some(v)` if the given key was previously mapped to some value `v`, or `None` otherwise
    */
  def replace(k: K, v: V): Option[V]

  override def getOrElseUpdate(key: K, @deprecatedName("op", since="2.13.13") defaultValue: => V): V = get(key) match {
    case Some(v) => v
    case None =>
      val v = defaultValue
      putIfAbsent(key, v) match {
        case Some(ov) => ov
        case None => v
      }
  }
```

2.12
```scala
  /** If given key is already in this map, returns associated value.
   *
   *  Otherwise, computes value from given expression `op`, stores with key
   *  in map and returns that value.
   *
   *  Concurrent map implementations may evaluate the expression `op`
   *  multiple times, or may evaluate `op` without inserting the result.
   *  
   *  @param  key the key to test
   *  @param  op  the computation yielding the value to associate with `key`, if
   *              `key` is previously unbound.
   *  @return     the value associated with key (either previously or as a result
   *              of executing the method).
   */
  def getOrElseUpdate(key: K, op: => V): V =
    get(key) match {
      case Some(v) => v
      case None => val d = op; this(key) = d; d
    }

```

Reply:
@frankvicky we should check the `JavaCollectionWrappers` for 2.13

https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/convert/JavaCollectionWrappers.scala#L482

```scala
    override def getOrElseUpdate(key: K, op: => V): V =
      underlying.computeIfAbsent(key, _ => op) match {
        case null => super/*[concurrent.Map]*/.getOrElseUpdate(key, op)
        case v    => v
      }
```

Hence, the wrapper actually leverage `computeIfAbsent` for normal case. The `getOrElseUpdate` is used to put the `null` to map since `computeIfAbsent` does not. That is for following `getOrElseUpdate` semantics

The `getOrElseUpdate` of current scala 2.13 is good, and we can do a bit cleanup (in the follow-up) as current usages (`markersQueuePerBroker` and `markersPerTxnTopicPartition`) are not necessary to use scala wrapper ..

Reply:
`computeIfAbsent` acquires locks, so the behavior is different.

Reply:
> The Java implementation of computeIfAbsent acquires locks, so the behavior is different.

Pardon me, either `computeIfAbsent` or `putIfAbsent` acquires lock (`ConcurrentHashMap` impl) when it needs to mutable the inner table. Additionally, the main behavior difference is that `atomicGetOrUpdate` has a side effect that it may create unnecessary value in race condition, but `getOrElseUpdate`  (`computeIfAbsent`) does not. For another, the main concern in this use case is the optimized read (according to comment). Maybe we can run JMH to check the perf of read-heavy case between `computeIfAbsent` and `get`. 

@frankvicky Could you please add a test case to `ConcurrentMapBenchmark`?

Reply:
Test in macOSX
```
Benchmark                                                            (mapSize)  (writePercentage)  Mode  Cnt  Score   Error  Units
ConcurrentMapBenchmark.testConcurrentHashMapComputeIfAbsentReadOnly        100                0.1  avgt    5  4.706 ± 0.074  ns/op
ConcurrentMapBenchmark.testConcurrentHashMapGetReadOnly                    100                0.1  avgt    5  3.557 ± 0.032  ns/op
JMH benchmarks done
```

Reply:
@chia7712 Right, one implementation _always_ acquires a lock to ensure the value is only computed once while the other only acquires locks if it needs to mutate the underlying map. The latter is better for read-heavy while the former is better for write-heavy cases. I think it would be best to have changes that may make things worse submitted separately. This PR should only have the changes that are neutral or always better.

Reply:
>  I think it would be best to have changes that may make things worse submitted separately. This PR should only have the changes that are neutral or always better.

totally agree. open https://issues.apache.org/jira/browse/KAFKA-17677 as follow-up. @frankvicky could you please keep `atomicGetOrUpdate`?